File: packages/ui-swiftui/Sources/KinetixUI/NotificationCenter.swift
//
// NotificationCenter.swift — KinetixNotificationCenter family.
//
// Mirrors packages/ui/src/components/notification-center.tsx: a bell
// trigger opening a popover list of read/unread items with a "mark all
// read" action. Built directly on SwiftUI's `.popover` (same as
// KinetixPopover) rather than a hand-rolled overlay. Unlike the Compose/
// Flutter ports, SF Symbols give a real system bell glyph — no "no icon
// library" gap to document here.
//

import SwiftUI

public struct KinetixNotificationCenterTrigger: View {
    @Environment(\.kinetixColors) private var colors
    private let unreadCount: Int
    private let action: () -> Void

    public init(unreadCount: Int = 0, action: @escaping () -> Void) {
        self.unreadCount = unreadCount
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            Image(systemName: "bell")
                .font(.system(size: 18))
                .foregroundStyle(colors.foreground)
                .frame(width: 36, height: 36)
        }
        .buttonStyle(.plain)
        .overlay(alignment: .topTrailing) {
            if unreadCount > 0 {
                Circle()
                    .fill(colors.destructive)
                    .frame(width: 8, height: 8)
                    .overlay(Circle().strokeBorder(colors.background, lineWidth: 2))
            }
        }
    }
}

public struct KinetixNotificationCenter<Anchor: View, Content: View>: View {
    @Environment(\.kinetixColors) private var colors

    @Binding private var isPresented: Bool
    private let onMarkAllRead: (() -> Void)?
    private let anchor: Anchor
    private let content: Content

    public init(
        isPresented: Binding<Bool>,
        onMarkAllRead: (() -> Void)? = nil,
        @ViewBuilder anchor: () -> Anchor,
        @ViewBuilder content: () -> Content
    ) {
        self._isPresented = isPresented
        self.onMarkAllRead = onMarkAllRead
        self.anchor = anchor()
        self.content = content()
    }

    public var body: some View {
        anchor
            .popover(isPresented: $isPresented) {
                VStack(alignment: .leading, spacing: 0) {
                    HStack {
                        Text("Notifications")
                            .font(.kinetixTitleSm.weight(.medium))
                            .foregroundStyle(colors.foreground)
                        Spacer()
                        if let onMarkAllRead {
                            Button("Mark all read", action: onMarkAllRead)
                                .font(.kinetixLabelMd.weight(.medium))
                                .foregroundStyle(colors.primary)
                                .buttonStyle(.plain)
                        }
                    }
                    .padding(16) // spacing/4

                    ScrollView {
                        VStack(spacing: 0) { content }
                    }
                    .frame(maxHeight: 320)
                }
                .frame(minWidth: 320) // off the shared spacing scale
                .background(colors.popover)
            }
    }
}

public struct KinetixNotificationItem: View {
    @Environment(\.kinetixColors) private var colors

    private let title: String
    private let description: String?
    private let time: String?
    private let unread: Bool
    private let onSelect: (() -> Void)?

    public init(
        _ title: String,
        description: String? = nil,
        time: String? = nil,
        unread: Bool = false,
        onSelect: (() -> Void)? = nil
    ) {
        self.title = title
        self.description = description
        self.time = time
        self.unread = unread
        self.onSelect = onSelect
    }

    public var body: some View {
        Button(action: { onSelect?() }) {
            HStack(alignment: .top, spacing: 12) { // spacing/3
                Circle()
                    .fill(unread ? colors.primary : .clear)
                    .frame(width: 8, height: 8)
                    .padding(.top, 6)

                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(.kinetixBodySm)
                        .fontWeight(unread ? .medium : .regular)
                        .foregroundStyle(colors.foreground)
                        .lineLimit(1)
                    if let description {
                        Text(description)
                            .font(.kinetixBodySm)
                            .foregroundStyle(colors.mutedForeground)
                            .lineLimit(1)
                    }
                    if let time {
                        Text(time)
                            .font(.kinetixLabelSm)
                            .foregroundStyle(colors.mutedForeground)
                    }
                }
                Spacer(minLength: 0)
            }
            .padding(.horizontal, 16) // spacing/4
            .padding(.vertical, 12) // spacing/3
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(onSelect == nil)
    }
}

File: packages/ui-compose/ui/src/main/kotlin/com/kinetixui/ui/NotificationCenter.kt
package com.kinetixui.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.dimensionResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * KinetixNotificationCenter family — mirrors
 * `packages/ui/src/components/notification-center.tsx`: a bell trigger
 * opening a popover list of read/unread items with a "mark all read"
 * action. Built directly on [KinetixDropdownMenu] — the same "it's
 * already the thing we built" reuse [KinetixSelect] made — with the
 * caller owning `expanded`/`onDismissRequest`, same convention as every
 * other anchored-popover composable here. No icon library wired in yet
 * (same documented gap as [KinetixInform]) — the bell is a plain glyph.
 */
@Composable
fun KinetixNotificationCenterTrigger(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    unreadCount: Int = 0,
) {
    val colors = KinetixColorScheme.current
    Box(modifier = modifier) {
        KinetixButton(onClick, variant = KinetixButtonVariant.Ghost, size = KinetixButtonSize.Icon) {
            Text("🔔", fontSize = 18.sp)
        }
        if (unreadCount > 0) {
            Box(
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .padding(6.dp)
                    .size(8.dp)
                    .border(2.dp, colors.background, CircleShape)
                    .background(colors.destructive, CircleShape),
            )
        }
    }
}

@Composable
fun KinetixNotificationCenter(
    expanded: Boolean,
    onDismissRequest: () -> Unit,
    modifier: Modifier = Modifier,
    onMarkAllRead: (() -> Unit)? = null,
    trigger: @Composable () -> Unit,
    content: @Composable ColumnScope.() -> Unit,
) {
    val colors = KinetixColorScheme.current

    KinetixDropdownMenu(visible = expanded, onDismissRequest = onDismissRequest, modifier = modifier, anchor = trigger) {
        Column(modifier = Modifier.widthIn(min = 320.dp)) {
            Row(
                modifier = Modifier
                    .padding(horizontal = dimensionResource(R.dimen.spacing_4), vertical = dimensionResource(R.dimen.spacing_3)),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    text = "Notifications",
                    color = colors.foreground,
                    fontWeight = FontWeight.Medium,
                    fontSize = dimensionResource(R.dimen.font_size_title_sm).value.sp,
                )
                if (onMarkAllRead != null) {
                    Text(
                        text = "Mark all read",
                        color = colors.primary,
                        fontWeight = FontWeight.Medium,
                        fontSize = dimensionResource(R.dimen.font_size_label_md).value.sp,
                        modifier = Modifier.clickable(onClick = onMarkAllRead),
                    )
                }
            }
            Column(modifier = Modifier.heightIn(max = 320.dp), content = content)
        }
    }
}

@Composable
fun KinetixNotificationItem(
    title: String,
    modifier: Modifier = Modifier,
    description: String? = null,
    time: String? = null,
    unread: Boolean = false,
    onSelect: (() -> Unit)? = null,
) {
    val colors = KinetixColorScheme.current
    Row(
        modifier = modifier
            .clickable(enabled = onSelect != null, onClick = { onSelect?.invoke() })
            .padding(horizontal = dimensionResource(R.dimen.spacing_4), vertical = dimensionResource(R.dimen.spacing_3)),
        horizontalArrangement = Arrangement.spacedBy(dimensionResource(R.dimen.spacing_3)),
    ) {
        Box(
            modifier = Modifier
                .padding(top = 6.dp)
                .size(8.dp)
                .background(if (unread) colors.primary else Color.Transparent, CircleShape),
        )
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = title,
                color = colors.foreground,
                fontWeight = if (unread) FontWeight.Medium else FontWeight.Normal,
                fontSize = dimensionResource(R.dimen.font_size_body_sm).value.sp,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            if (description != null) {
                Text(
                    text = description,
                    color = colors.mutedForeground,
                    fontSize = dimensionResource(R.dimen.font_size_body_sm).value.sp,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
            if (time != null) {
                Text(
                    text = time,
                    color = colors.mutedForeground,
                    fontSize = dimensionResource(R.dimen.font_size_label_sm).value.sp,
                )
            }
        }
    }
}
